sheets/google_sheets/upload_file.py

```diff
@@ -0,0 +1,263 @@
+from google.oauth2 import service_account
+
+SCOPES = ['https://spreadsheets.google.com/feeds',
+          'https://www.googleapis.com/auth/drive']
+socket.setdefaulttimeout(600)
+#SCOPES = ['https://www.googleapis.com/auth/spreadsheets']
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--source-file-name', dest='source_file_name',
+            required=True)
+    parser.add_argument('--source-folder-name', dest='source_folder_name',
+            default='', required=False)
+    parser.add_argument('--destination-file-name',
+            dest='file_name', default='', required=False)
+    parser.add_argument('--starting-cell',
+            dest='starting_cell', default='A1', required=False)
+    parser.add_argument('--tab-name', dest='tab_name',
+            default=None, required=False)
+    parser.add_argument('--service-account', dest='gcp_application_credentials',
+            default=None, required=True)
+    return parser.parse_args()
+
+
+def set_environment_variables(args):
+    """
+    Set GCP credentials as environment variables if they're provided via keyword
+    arguments rather than seeded as environment variables. This will override
+    system defaults.
+    """
+    credentials = args.gcp_application_credentials
+    try:
+        json_credentials = json.loads(credentials)
+        fd, path = tempfile.mkstemp()
+        print(f'Storing json credentials temporarily at {path}')
+        with os.fdopen(fd, 'w') as tmp:
+            tmp.write(credentials)
+        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = path
+        return path
+    except Exception:
+        print('Using specified json credentials file')
+        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = credentials
+        return
+
+
+def clean_folder_name(folder_name):
+    """
+    Cleans folders name by removing duplicate '/' as well as leading and
+    trailing '/' characters.
+    """
+    folder_name = folder_name.strip('/')
+    if folder_name != '':
+        folder_name = os.path.normpath(folder_name)
+    return folder_name
+
+
+def combine_folder_and_file_name(folder_name, file_name):
+    """
+    Combine together the provided folder_name and file_name into one path
+    variable.
+    """
+    combined_name = os.path.normpath(
+        f'{folder_name}{"/" if folder_name else ""}{file_name}')
+    combined_name = os.path.normpath(combined_name)
+
+    return combined_name
+
+
+def check_workbook_exists(service, spreadsheet_id, tab_name):
+    """
+    Checks if the workbook exists within the spreadsheet.
+    """
+    try:
+        spreadsheet = service.spreadsheets().get(spreadsheetId=spreadsheet_id).execute()
+        sheets = spreadsheet['sheets']
+        exists = [True for sheet in sheets if sheet['properties']['title'] == tab_name]
+        return True if exists else False
+    except Exception as e:
+        print(f'Failed to check workbook {tab_name} for spreadsheet ' \
+                f'{spreadsheet_id}')
```

Comment:
These dynamic variables should be flipped and contain Google Sheet syntax. I got a connection error that confused me here.

'Failed to check spreadsheet {spreadsheet_id} for a sheet named {tab_name}`